libimage/manifests/manifests.go

```diff
@@ -455,6 +634,222 @@ func (l *list) Add(ctx context.Context, sys *types.SystemContext, ref types.Imag
+		}
+		subject = &v1.Descriptor{
+			MediaType: subjectManifestType,
+			Digest:    subjectManifestDigest,
+			Size:      int64(len(subjectManifestBytes)),
+		}
+		l.instances[subjectManifestDigest] = transports.ImageName(options.SubjectReference)
+	}
+
+	// Build up the layers list piece by piece.
+	var layers []v1.Descriptor
+	fileDigests := make(map[string]digest.Digest)
+
+	if len(files) == 0 {
+		// https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidelines-for-artifact-usage
+		// says that we should have at least one layer listed, even if it's just a placeholder
+		layers = append(layers, v1.DescriptorEmptyJSON)
+	}
+	for _, file := range files {
+		if err := func() error {
+			// Open the file so that we can digest it.
+			absFile, err := filepath.Abs(file)
+			if err != nil {
+				return fmt.Errorf("converting %q to an absolute path: %w", file, err)
+			}
+
+			f, err := os.Open(absFile)
+			if err != nil {
+				return fmt.Errorf("reading %q to determine its digest: %w", file, err)
+			}
+			defer f.Close()
+
+			// Hang on to a copy of the first 512 bytes, but digest the whole thing.
+			digester := digest.Canonical.Digester()
+			writeCounter := ioutils.NewWriteCounter(digester.Hash())
+			var detectableData bytes.Buffer
+			_, err = io.CopyN(writeCounter, io.TeeReader(f, &detectableData), 512)
+			if err != nil && !errors.Is(err, io.EOF) {
+				return fmt.Errorf("reading %q to determine its digest: %w", file, err)
+			}
+			if err == nil {
+				if _, err := io.Copy(writeCounter, f); err != nil {
+					return fmt.Errorf("reading %q to determine its digest: %w", file, err)
+				}
+			}
+			fileDigests[absFile] = digester.Digest()
+
+			// If one wasn't specified, figure out what the MediaType should be.
+			title := filepath.Base(absFile)
+			layerMediaType := options.LayerMediaType
+			if layerMediaType == nil {
+				if index := strings.LastIndex(title, "."); index != -1 {
+					// File's basename has an extension, try to use a shortcut.
+					tmp := mime.TypeByExtension(title[index:])
+					if tmp != "" {
+						layerMediaType = &tmp
+					}
+				}
+				if layerMediaType == nil {
+					// File's basename has no extension or didn't map to a type, look at the contents we saved.
+					tmp := http.DetectContentType(detectableData.Bytes())
+					layerMediaType = &tmp
+				}
+				if layerMediaType != nil {
+					// Strip off any parameters, since we only want the type name.
+					if parsedMediaType, _, err := mime.ParseMediaType(*layerMediaType); err == nil {
+						layerMediaType = &parsedMediaType
+					}
+				}
+			}
+
+			// Only set an ArtifactType for the layer if one was specified.
+			layerArtifactType := options.LayerArtifactType
+			if layerArtifactType == nil {
+				tmp := ""
+				layerArtifactType = &tmp
+			}
+			descriptor := v1.Descriptor{
+				MediaType:    *layerMediaType,
+				ArtifactType: *layerArtifactType,
```

Comment:
I can’t see anything in the OCI guidance about layer descriptors having this field, and I don’t know what the semantics should be.

That this is even possible seems to be an artifact of the imprecise type design, reusing the same `Descriptor` in various situations where different subsets are relevant.

Reply:
Yeah, this is a hedge.  If you think we're not likely to need it, I have no problem with removing it.